lib/site-inspector.rb

```diff
@@ -127,37 +134,213 @@ def to_json
+    details[:hsts_entire_domain] = (
+      combos[:https][:root][:hsts] and
+      combos[:https][:root][:hsts_header].downcase.include?("includesubdomains")
+    )
+
+    # HSTS preload --ready?
+    details[:hsts_entire_domain_preload] = (
+      details[:hsts_entire_domain] and
+      combos[:https][:root][:hsts_header].downcase.include?("preload")
+    )
+
+    details
+  end
+
+  def endpoints
+    https_www = http_endpoint(true, true)
+    http_www = http_endpoint(false, true)
+    https_root = http_endpoint(true, false)
+    http_root = http_endpoint(false, false)
+
     {
-      :domain => domain.to_s,
-      :uri => uri.to_s,
-      :government => government?,
-      :live => !!response,
-      :ssl => https?,
-      :enforce_https => enforce_https?,
-      :non_www => non_www?,
-      :redirect => redirect,
-      :ip => ip,
-      :hostname => hostname.to_s,
-      :ipv6 => ipv6?,
-      :dnssec => dnssec?,
-      :cdn => cdn,
-      :google_apps => google_apps?,
-      :cloud_provider => cloud_provider,
-      :server => server,
-      :cms => cms,
-      :analytics => analytics,
-      :javascript => javascript,
-      :advertising => advertising,
-      :slash_data => slash_data?,
-      :slash_developer => slash_developer?,
-      :data_dot_json => data_dot_json?,
-      :click_jacking_protection => click_jacking_protection?,
-      :content_security_policy => content_security_policy?,
-      :xss_protection => xss_protection?,
-      :secure_cookies => secure_cookies?,
-      :strict_transport_security => strict_transport_security?,
-      :headers => headers
+      https: {
+        www: https_www,
+        root: https_root
+      },
+      http: {
+        www: http_www,
+        root: http_root
+      }
     }
   end
+
+  # State of affairs at a particular endpoint.
+  def http_endpoint(ssl, www)
+    details = {}
+
+    # Don't follow redirects for first ping.
+    response = request(ssl, www, false)
+
+
+    # For HTTPS: examine the full range of possibilities.
+    if ssl
+      if response.return_code == :ok
+        details[:https_valid] = true
+
+      # Bad certificate chain.
+      elsif response.return_code == :ssl_cacert
+        details[:https_valid] = false
+        details[:https_bad_chain] = true
+        response = request(ssl, www, false, false, true)
+        # Bad everything.
+        if response.return_code == :peer_failed_verification
+          details[:https_bad_name] = true
+          response = request(ssl, www, false, false, false)
+        end
+      # Bad hostname.
+      elsif response.return_code == :peer_failed_verification
+        details[:https_valid] = false
+        details[:https_bad_name] = true
+        response = request(ssl, www, false, true, false)
+        # Bad everything.
+        if response.return_code == :ssl_cacert
+          details[:https_bad_chain] = true
+          response = request(ssl, www, false, false, false)
+        end
+      end
+    end
```

Comment:
The point of the above section is to end up with an understanding of an HTTPS endpoint before declaring it "broken", using Typhoeus/libcurl's return codes. This discovers whether a domain is at least listening over port 443, and whether it has an invalid hostname, an untrusted certificate chain, or both. 

Those values are stored in `details` as this executes, and they are used again once below, to successfully follow redirects and discover whether the endpoint eventually leads to a 200 response (whether the cert is trusted or not).